src/cascadia/Remoting/WindowManager.cpp

```diff
@@ -801,4 +801,22 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
         }
         return nullptr;
     }
+
+    // Attempt to summon an existing window. This static function does NOT
+    // pre-register as the monarch. This is used for activations from a
+    // notification, where this process should NEVER become its own window.
+    bool WindowManager::SummonForNotification(const uint64_t windowId)

```

Comment:
I'm not sure what exactly is expected of this, but it doesn't seem to focus the Terminal window for me. If I've minimized it, it will be restored, but it's still hidden behind whatever other applications I've got open.

Also, if I've switched to another Terminal tab, I would have expected it to switch to the tab that triggered the notification, but that doesn't seem to be the case.

If that's not expected to work, though, that's fine. Just wanted to let you know in case there is a bug here.

Reply:
> it's still hidden behind whatever other applications I've got open

This is based off the same `globalSummon` code, which doesn't always work depending on the FG application. I know for a fact that Task Manager and OneNote don't allow you to summon the window on top of them - what did you have focused?

> if I've switched to another Terminal tab, I would have expected it to switch to the tab that triggered the notification

That's something we should definitely do! I'd probably hold that for a follow-up though, to minimize the diffs. Right now I'm not really passing the `tabIndex` parameter, and there's not a good way to "`globalSummon` and switch to a tab" all in one go.

Reply:
> what did you have focused?

I've tried a few things now - Notepad, Firefox, Calculator - haven't yet found anything that worked. Even when I can see it being restored from a minimized state, it restores itself behind the active window.

Maybe it's just my version of Windows? (10.0.19044.2130)

Reply:
huh. That's... super weird. Does `globalSummon` work for you in this scenario?

We do play _weird games_ with inserting a thread into the FG process so we can activate it (unsurprisingly, Windows does not make it easy for an app to bring itself to the FG). But if this didn't work, then I'd assume that `globalSummon` wouldn't work the same....

Unless there's something extra weird where the FG application is like, `explorer.exe` itself... 

I'll spin this up in a VM to check.

Reply:
> Does `globalSummon` work for you in this scenario?

Yeah. That seems to work in most cases. If it's minimized it pops up over the foreground app, and if it's open but in the background, it'll be moved to the foreground.

Although the latter case doesn't seem work with UWP apps I think (I've tried Calculator, Skype, News). If Terminal is hidden/minimized, global summon will open it up in the foreground, but if it's already open, and I've got something like Calculator focused, Terminal doesn't seem able to move itself to the foreground.

Reply:
Interesting. Finally got a chance to try this in the 10 VM. Seems like the notification will restore the Terminal from minimized, but won't bring it to the foreground. If it's open (but not in the FG), then this does seemingly nothing. Interesting. 

~I think this has to do with the way `globalSummon` attempts to bring a window to the foreground, but it interacts with the Windows 10 notifications in a VERY weird way. Usually to summon, we do this weird "inject a thread, then SetForeground on that thread" trick so that we~
~I _suspect_ that when we get invoked from the Windows 10 notification, we can't~

Now that I've typed that up, that can't be right. 

In `IslandWindow::_globalActivateWindow`,  
* if minimized, we `ShowWindow(..., SW_RESTORE);`, then `SetWindowPos(..., SWP_NOZORDER | SWP_NOSIZE | SWP_NOACTIVATE)` (to put on the right monitor). We're NOT doing the wacky FG trick here. 
* if _not_ minimized, we `AttachThreadInput`+`ShowWindow`+`SetActiveWindow`+`SetWindowPos` to take foreground, activate ourselves, and move to the right monitor. 

However, in neither of these cases does the Terminal get brought to the foreground from a notification in Windows 10. I'd think that the restore one _should_. Maybe there's a bug in Windows 10? Lemme reach out to a notifications expert.

Reply:
A thought - Outlook and Teams notifications _never_ open the window in the foreground when you click on their notifications (certainly not on Windows 10). That's weird, but maybe a lead.

Reply:
Could it perhaps have something to do with Windows trying to prevent focus stealing? I don't know anything about the architecture, so maybe this doesn't apply, but could it help if you called [AllowSetForegroundWindow](https://learn.microsoft.com/en-gb/windows/win32/api/winuser/nf-winuser-allowsetforegroundwindow) somewhere?

Reply:
Almost certainly. 

You know, that's not a bad thought. That might require some wacky plumbing, but that might work. We'd need to get the PID of the target window to the WindowsTerminal.exe that got spawned, _before_ we do then windowManager activation.... Might be possible? Certainly will be easier after #14843